### PR TITLE
add FastNetMessages implementation of SayText, SayText2 and ServerRan…

### PR DESCRIPTION
**Pull request by akiver**
…kUpdate

If SLOW_PROTOBUF is used, this events will not be triggered and you will have to use the old method to get them.

ref #95 #96

### CODE REVIEW

**Comment by akiver** [comment]
Hi, I tested it on multiple demos from Valve, eBot and Faceit without errors.
SayText events aren't triggered as conversations aren't networked.

I'm not sure about how I should raise this kind of messages as there is no descriptors like GameEventList.
So I decided to raise it just after the end of parsing.

Do you think that fields EntityIndex, Chat and ChatAllText are necessary as they are everytime equals to 0 or false (SayText)?

File: DemoInfo/DP/FastNetmessages/SayText2.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+		private void Raise(DemoParser parser)
+		{
+			Player sender = null;
+			// Find the sender
+			foreach (KeyValuePair<int, Player> keyValuePair in parser.Players)
+			{
+				if (keyValuePair.Value.Name == Params[0])
+				{
+					sender = parser.Players[keyValuePair.Key];
```
Comment (main--) [inline DemoInfo/DP/FastNetmessages/SayText2.cs]: Why are you doing the lookup here? Shouldn't `keyValuePair.Value` already be exactly the element you want?

Comment (main--) [inline DemoInfo/DP/FastNetmessages/SayText2.cs]: Now that I spent some time thinking about this, how about:

```csharp
var sender = parser.Players.Values.FirstOrDefault(x => x.Name == Params[0]);
```

That should do the job just fine, right?

Besides, is there any alternative to matching on the name here? That `EntIdx` field looks like it might allow association to player entities.

Comment (akiver) [inline DemoInfo/DP/FastNetmessages/SayText2.cs]: It should yes, I will test and update it.
I didn't find an other way, EntIdx doesn't match players id :/

**Comment by main--** [comment]
Thank you very much for this contribution.

>Hi, I tested it on multiple demos from Valve, eBot and Faceit without errors.

Awesome!

>I'm not sure about how I should raise this kind of messages as there is no descriptors like GameEventList.
>So I decided to raise it just after the end of parsing.

The way you did it is fine. The only note I'd add is that it might make sense to remove `ServerRankUpdateEventArgs` and instead just fire the event for every individual `RankUpdateEventArgs` this is not a big deal.

>Do you think that fields EntityIndex, Chat and ChatAllText are necessary as they are everytime equals to 0 or false (SayText)?

Generally, we tried to keep the FastNetmessages as close to the protobuf definitions as possible, so you should definitely keep them there.

As for the events - our general approach is that we only pass data out of demoinfo that is actually "useful". I'm not familiar enough with the corresponding source engine implementation details (although @moritzuehling probably is), so I'm not sure about this but if there really is no way for these fields to ever have different values (or if we just have absolutely no idea what that would mean), I don't think they should be passed on through the event.

---

Thanks for sticking with my review. Other potential improvements I've noticed:

* All of our existing FastNetmessages are `struct`s instead of `class`es to avoid the allocation overhead. It's small but it adds up. So please adjust your new ones accordingly.
* While we strongly advise against using `SLOW_PROTOBUF` mode in production environments, it's still a very useful debugging aid. Some of the events and methods you added conflict with the ones you added in #65 and #70, causing compile errors. Now that we have the fast implementations, you can just remove the old code (a simple `git revert` should do the job). I did notice though that your old versions have quite extensive doc comments while the new ones don't - keeping those would obviously be great.

---

Regarding the [mysterious entity index](https://github.com/EHVAG/demoinfo/pull/97#discussion_r55923416): @moritzuehling any idea on this?

**Comment by akiver** [comment]
As I added this [check](https://github.com/EHVAG/demoinfo/pull/97/files#diff-c3ec13004e53065fce5b84e867fadb2bR50), should I remove all protobuf related methods such as this [one](https://github.com/akiver/demoinfo/blob/usermessages/DemoInfo/DP/Handler/UserMessageHandler.cs#L31) or I left them but enable the new implementation only when the SLOW_PROTOBUF flag isn't present?

My thought is, if SLOW_PROTOBUF is present, don't use the new fastnetmessages and keep the old protobuf parsing methods otherwise use only the new system. What's your pov?

**Comment by main--** [comment]
>My thought is, if SLOW_PROTOBUF is present, don't use the new fastnetmessages and keep the old protobuf parsing methods otherwise use only the new system. What's your pov?

That would be a possible approach, but it's not what we've been doing in the past. `SLOW_PROTOBUF` is intended for debugging, not just in the sense of a quick check but in the sense of a thorough investigation that involves writing code. It was not designed as a backup/fallback (although you can of course temporarily replace fastnetmessages with slow protobuf code if that helps with the specific issue you're debugging) but rather (in its "default configuration") as a way to investigate unimplemented messages when you're looking for pieces of information DemoInfo doesn't yet offer. Your `SLOW_PROTOBUF`-based contributions were a slight deviation from this path, but of course we're happy about every bit of help.

So to sum things up, I'd recommend that you just remove all of the old code as we now have a faster replacement that works reliably according to your observations.

**Comment by akiver** [comment]
Thanks, I made changes according to your feedbacks.

File: DemoInfo/DP/DemoPacketParser.cs [hunk 2]
```diff
@@ -45,7 +48,13 @@ public static void ParsePacket(IBitStream bitstream, DemoParser demo)
 					new UpdateStringTable().Parse(bitstream, demo);
 				} else if (cmd == (int)NET_Messages.net_Tick) { //and all this other stuff
 						new NETTick().Parse(bitstream, demo);
-				} else {
+				}
+				#if !SLOW_PROTOBUF
```
Comment (main--) [inline DemoInfo/DP/DemoPacketParser.cs]: I guess these checks are no longer appropriate now that you've removed the alternate code.

**Comment by main--** [comment]
Thanks a lot, let's wait a day or two for @moritzuehling to have a look at this and we'll soon be ready to merge! :smile:
